contracts/cw20-ics20/src/ibc.rs

```diff
@@ -338,6 +336,9 @@ fn on_packet_failure(
 ) -> Result<IbcBasicResponse, ContractError> {
     let msg: Ics20Packet = from_binary(&packet.data)?;
 
+    // undo the balance update
+    reduce_channel_balance(deps.storage, &packet.src.channel_id, &msg.denom, msg.amount)?;
```

Comment:
My concern would be: What if, for whatever reason, the packet failure message never arrives?

Not sure if this can be abused, as the successful deposit on the other side is the one crediting the funds. But still, if the packet failure never arrives, the balance would be left in an inconsistent state, resulting in lost funds.

Reply:
It should eventually arrive. If nothing out via "timeout".

If this never gets relayed, the senders tokens remained locked in the contract. Once the timeout arrives, they are paid back.